stdlib/runtime/numbers.gr

```diff
@@ -853,69 +1015,207 @@ let numberAddSubInt32Help = (xval, y, isSub) => {
 }
 
 @unsafe
-let rec numberAddSubHelp = (x, y, isSub) => {
-  if (isSimpleNumber(x)) {
-    numberAddSubSimpleHelp(x, y, isSub)
+let numberAddSubBigIntHelp = (x, y, isSub) => {
+  if (isSimpleNumber(y)) {
+    let yval = WasmI64.extendI32S(untagSimple(y))
+    let yBig = BI.makeWrappedInt64(yval)
+    let res = if (isSub) {
+      BI.sub(x, yBig)
+    } else {
+      BI.add(x, yBig)
+    }
+    Memory.decRef(yBig)
+    reducedBigInteger(res)
   } else {
-    let xtag = boxedNumberTag(x)
-    match (xtag) {
+    let yBoxedNumberTag = boxedNumberTag(y)
+    match (yBoxedNumberTag) {
       t when WasmI32.eq(t, Tags._GRAIN_INT32_BOXED_NUM_TAG) => {
-        numberAddSubInt32Help(boxedInt32Number(x), y, isSub)
+        let yBoxedVal = WasmI64.extendI32S(boxedInt32Number(y))
+        let yBig = BI.makeWrappedInt64(yBoxedVal)
+        let res = if (isSub) {
+          BI.sub(x, yBig)
+        } else {
+          BI.add(x, yBig)
+        }
+        Memory.decRef(yBig)
+        reducedBigInteger(res)
       },
       t when WasmI32.eq(t, Tags._GRAIN_INT64_BOXED_NUM_TAG) => {
-        numberAddSubInt64Help(boxedInt64Number(x), y, isSub)
+        let yBoxedVal = boxedInt64Number(y)
+        let yBig = BI.makeWrappedInt64(yBoxedVal)
+        let res = if (isSub) {
+          BI.sub(x, yBig)
+        } else {
+          BI.add(x, yBig)
+        }
+        Memory.decRef(yBig)
+        reducedBigInteger(res)
+      },
+      t when WasmI32.eq(t, Tags._GRAIN_BIGINT_BOXED_NUM_TAG) => {
+        let res = if (isSub) {
+          BI.sub(x, y)
+        } else {
+          BI.add(x, y)
+        }
+        reducedBigInteger(res)
       },
       t when WasmI32.eq(t, Tags._GRAIN_RATIONAL_BOXED_NUM_TAG) => {
-        numberAddSubRationalHelp(x, y, isSub)
+        let yNumerator = boxedRationalNumerator(y)
+        let yDenominator = boxedRationalDenominator(y)
+        let expandedXNumerator = BI.mul(x, yDenominator)
+        let result = if (isSub) BI.sub(expandedXNumerator, yNumerator)
+          else BI.add(expandedXNumerator, yNumerator)
+        Memory.decRef(expandedXNumerator)
+        let ret = reducedFractionBigInt(result, yDenominator)
+        Memory.decRef(result)
+        ret
       },
       t when WasmI32.eq(t, Tags._GRAIN_FLOAT32_BOXED_NUM_TAG) => {
-        numberAddSubFloat32Help(boxedFloat32Number(x), y, isSub)
+        let xval = BI.toFloat32(x)
+        let yBoxedVal = boxedFloat32Number(y)
+        let result = if (isSub) WasmF32.sub(xval, yBoxedVal)
+          else WasmF32.add(xval, yBoxedVal)
+        // TODO: (#304) this isn't safe enough
```

Comment:
```suggestion
        // TODO(#304): this isn't safe enough
```

Reply:
There are some older TODOs in this file (and probably throughout the codebase) which don't align with this syntax. If you want to enforce a standard, someone should probably make a PR which cleans them all up.

Reply:
Sounds good - #1193